File: internal/app/commands/silly/capy.go
package silly

import (
	"encoding/json"
	"github.com/bwmarrin/discordgo"
	"meteor-bot/internal/app/common"
	"net/http"
)

type CapyCommand struct {
	common.BaseCommand
}

func NewCapyCommand() *CapyCommand {
	return &CapyCommand{
		BaseCommand: *common.NewCommandBuilder().
			SetName("capy").
			SetDescription("pulls up").
			Build(),
	}
}

func (c *CapyCommand) Build() *discordgo.ApplicationCommand {
	return &discordgo.ApplicationCommand{
		Name:        c.Name(),
		Description: c.Description(),
	}
}

func (c *CapyCommand) Handler() common.CommandHandler {
	return func(s *discordgo.Session, i *discordgo.InteractionCreate) {
		// Fetch the capybara image
		resp, err := http.Get("https://api.capy.lol/v1/capybara?json=true")
		if err != nil {
			c.HandleInteractionRespond(s, i, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "Failed to fetch capybara image",
					Flags:   discordgo.MessageFlagsEphemeral,
				},
			})
			return
		}
		defer resp.Body.Close()

		// Decode the response
		var jsonResponse map[string]any
		if err = json.NewDecoder(resp.Body).Decode(&jsonResponse); err != nil {
			c.HandleInteractionRespond(s, i, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "Failed to decode the response",
					Flags:   discordgo.MessageFlagsEphemeral,
				},
			})
			return
		}

		// Extract the image URL
		data, ok := jsonResponse["data"].(map[string]any)
		if !ok {
			c.HandleInteractionRespond(s, i, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "Failed to parse the response",
					Flags:   discordgo.MessageFlagsEphemeral,
				},
			})
			return
		}
		url, ok := data["url"].(string)
		if !ok {
			c.HandleInteractionRespond(s, i, &discordgo.InteractionResponse{
				Type: discordgo.InteractionResponseChannelMessageWithSource,
				Data: &discordgo.InteractionResponseData{
					Content: "Failed to parse the response",
					Flags:   discordgo.MessageFlagsEphemeral,
				},
			})
			return
		}

		// Respond to the interaction
		c.HandleInteractionRespond(s, i, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: url,
			},
		})
	}
}

File: src/commands/silly/capy.rs
use reqwest;
use serde_json;
use serenity::builder::{
    CreateCommand, CreateInteractionResponse, CreateInteractionResponseMessage,
};
use serenity::model::application::InteractionResponseFlags;
use serenity::model::application::ResolvedOption;

pub fn register() -> CreateCommand {
    CreateCommand::new("capybara").description("pulls up")
}

pub async fn handle(_options: &[ResolvedOption<'_>]) -> CreateInteractionResponse {
    let api_url = "https://api.capy.lol/v1/capybara?json=true";
    let resp = match reqwest::get(api_url).await {
        Ok(r) => r,
        Err(_) => {
            let data = CreateInteractionResponseMessage::new()
                .content("Failed to fetch capybara image")
                .flags(InteractionResponseFlags::EPHEMERAL);
            return CreateInteractionResponse::Message(data);
        }
    };

    let json: serde_json::Value = match resp.json().await {
        Ok(j) => j,
        Err(_) => {
            let data = CreateInteractionResponseMessage::new()
                .content("Failed to decode the response")
                .flags(InteractionResponseFlags::EPHEMERAL);
            return CreateInteractionResponse::Message(data);
        }
    };

    let url = json
        .get("data")
        .and_then(|d| d.get("url"))
        .and_then(|u| u.as_str());

    match url {
        Some(url) => {
            let data = CreateInteractionResponseMessage::new().content(url);
            CreateInteractionResponse::Message(data)
        }
        None => {
            let data = CreateInteractionResponseMessage::new()
                .content("Failed to parse the response")
                .flags(InteractionResponseFlags::EPHEMERAL);
            CreateInteractionResponse::Message(data)
        }
    }
}
